write.go

```diff
@@ -232,12 +242,173 @@ func writeProduceRequestV2(w *bufio.Writer, codec CompressionCodec, correlationI
 	}
 
-	return w.Flush()
+	buf := &bytes.Buffer{}
+	buf.Grow(int(size))
+	bufWriter := bufio.NewWriter(buf)
+
+	for _, msg := range msgs {
+		writeMessage(bufWriter, msg.Offset, attributes, msg.Time, msg.Key, msg.Value)
+	}
+	bufWriter.Flush()
+
+	return buf.Bytes(), nil
+}
+
+func writeRecordBatch(codec CompressionCodec, correlationID int32, clientId, topic string, partition int32, timeout time.Duration, requiredAcks int16, msgs ...Message) ([]byte, error) {
+	var size int32
+	size = 8 + // base offset
+		4 + // batch length
+		4 + // partition leader epoch
+		1 + // magic
+		4 + // crc
+		2 + // attributes
+		4 + // last offset delta
+		8 + // first timestamp
+		8 + // max timestamp
+		8 + // producer id
+		2 + // producer epoch
+		4 // base sequence
+
+	for _, msg := range msgs {
+		size += estimatedRecordSize(&msg)
+	}
+
+	buf := &bytes.Buffer{}
+	buf.Grow(int(size))
+	bufWriter := bufio.NewWriter(buf)
+
+	baseTime := msgs[0].Time
+
+	baseOffset := baseOffset(msgs...)
+
+	writeInt64(bufWriter, baseOffset)
+
+	remainderBuf := &bytes.Buffer{}
+	remainderBuf.Grow(int(size - 12)) // 12 = batch length + base offset sizes
+	remainderWriter := bufio.NewWriter(remainderBuf)
+
+	writeInt32(remainderWriter, -1) // partition leader epoch
+	writeInt8(remainderWriter, 2)   // magic byte
+
+	//writeInt16(remainderWriter, 0) // TODO write crc
+
+	crcBuf := &bytes.Buffer{}
+	crcBuf.Grow(int(size - 12)) // 12 = batch length + base offset sizes
+	crcWriter := bufio.NewWriter(crcBuf)
+
+	writeInt16(crcWriter, 0) // attributes no compression, timestamp type 0 - create time, not part of a transaction, no control messages
+	writeInt32(crcWriter, int32(msgs[len(msgs)-1].Offset-baseOffset))
+	writeInt64(crcWriter, timestamp(baseTime))
+	lastTime := timestamp(msgs[len(msgs)-1].Time)
+	writeInt64(crcWriter, int64(lastTime))
+	writeInt64(crcWriter, -1)               // default producer id for now
+	writeInt16(crcWriter, -1)               // default producer epoch for now
+	writeInt32(crcWriter, -1)               // default base sequence
+	writeInt32(crcWriter, int32(len(msgs))) // record count
+
+	for _, msg := range msgs {
+		writeRecord(crcWriter, CompressionNoneCode, baseTime, baseOffset, msg)
+	}
+	crcWriter.Flush()
+
+	crcTable := crc32.MakeTable(crc32.Castagnoli)
+	crcChecksum := crc32.Checksum(crcBuf.Bytes(), crcTable)
+
+	writeInt32(remainderWriter, int32(crcChecksum))
+	remainderWriter.Write(crcBuf.Bytes())
+
+	remainderWriter.Flush()
+
+	writeInt32(bufWriter, int32(remainderBuf.Len()))
+	bufWriter.Write(remainderBuf.Bytes())
+	bufWriter.Flush()
+
+	return buf.Bytes(), nil
+}
+
+var maxDate time.Time = time.Date(5000, time.January, 0, 0, 0, 0, 0, time.UTC)
+
+func baseTime(msgs ...Message) (baseTime time.Time) {
+	baseTime = maxDate
+	for _, msg := range msgs {
+		if msg.Time.Before(baseTime) {
+			baseTime = msg.Time
+		}
+	}
+	return
+}
+
+const MaxUint = ^uint64(0)
+const MinUint = 0
+const MaxInt = int64(MaxUint >> 1)
+const MinInt = -MaxInt - 1
```

Comment:
Do we have to export those constants? 

Also why not use `math.MaxUint64`, `math.MaxInt64`, etc... ?

Reply:
Indeed. I should use standard values.